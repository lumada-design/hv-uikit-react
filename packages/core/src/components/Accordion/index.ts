export { default as accordionClasses } from "./accordionClasses";
export * from "./accordionClasses";
export * from "./Accordion";
