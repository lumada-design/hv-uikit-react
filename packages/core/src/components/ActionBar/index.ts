export { default as actionBarClasses } from "./actionBarClasses";
export * from "./actionBarClasses";
export * from "./ActionBar";
