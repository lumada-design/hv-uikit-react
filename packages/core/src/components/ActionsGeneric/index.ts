export { default as actionsGenericClasses } from "./actionsGenericClasses";
export * from "./actionsGenericClasses";
export * from "./ActionsGeneric";
