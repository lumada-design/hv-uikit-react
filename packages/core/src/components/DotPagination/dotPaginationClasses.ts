import { getClasses } from "@core/utils";

export type HvDotPaginationClasses = {
  root?: string;
  horizontal?: string;
  radioRoot?: string;
  radio?: string;
  icon?: string;
};

const classKeys: string[] = [
  "root",
  "horizontal",
  "radioRoot",
  "radio",
  "icon",
];

const dotPaginationClasses = getClasses<HvDotPaginationClasses>(
  classKeys,
  "HvDotPagination"
);

export default dotPaginationClasses;
