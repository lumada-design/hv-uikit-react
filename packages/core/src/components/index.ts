export * from "./Box";
export { Box as HvBox } from "./Box";
export * from "./Button";
export { Button as HvButton } from "./Button";
export * from "./Dropdown";
export { Dropdown as HvDropdown } from "./Dropdown";
export * from "./Header";
export {
  Actions as HvHeaderActions,
  Brand as HvHeaderBrand,
  Header as HvHeader,
  Navigation as HvHeaderNavigation,
} from "./Header";
export * from "./Typography";
export { Typography as HvTypography } from "./Typography";
