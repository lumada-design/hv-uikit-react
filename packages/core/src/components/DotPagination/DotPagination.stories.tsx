import styled from "@emotion/styled";
import {
  CurrentStep,
  RadioButtonUnselected,
} from "@hitachivantara/uikit-react-icons";
import { theme } from "@hitachivantara/uikit-styles";
import { Meta, StoryObj } from "@storybook/react";
import { HvTypography } from "@core/components";
import { useState } from "react";
import { HvDotPagination, HvDotPaginationProps } from "./DotPagination";
import dotPaginationClasses from "./dotPaginationClasses";

const meta: Meta<typeof HvDotPagination> = {
  title: "Components/Pagination/Dot Pagination",
  component: HvDotPagination,
};

export default meta;

export const Main: StoryObj<HvDotPaginationProps> = {
  argTypes: {
    classes: { control: { disable: true } },
  },
  render: () => {
    const [page, setPage] = useState<number>(0);
    const pages = [
      "This is page 1",
      "And this is page 2",
      "This is page 3",
      "This is page 4",
      "And finally, this is page 5",
    ];

    const StyledRoot = styled("div")({
      width: "100%",
      justifyContent: "center",
    });

    const StyledPage = styled("div")({
      textAlign: "center",
    });

    return (
      <StyledRoot>
        <StyledPage>
          <HvTypography>{pages[page]}</HvTypography>
        </StyledPage>
        <br />
        <HvDotPagination
          page={page}
          pages={pages.length}
          onPageChange={(_, value) => setPage(value)}
          getItemAriaLabel={(pageNumber) => {
            switch (pageNumber) {
              case 0:
                return "first page button aria-label";
              case 4:
                return "last page button aria-label";
              default:
                return `${pageNumber + 1} page aria-label`;
            }
          }}
          role="navigation"
          aria-label="Example Dot Navigation"
        />
      </StyledRoot>
    );
  },
};

export const CustomizedDotPagination: StoryObj<HvDotPaginationProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "The `DotPagination` component was customized to be similar to the NEXT Design System 3 specifications. " +
          "Thus, the styling was overridden and the `unselectedIcon` and `selectedIcon` properties were provided.",
      },
    },
  },
  render: () => {
    const [page, setPage] = useState<number>(0);
    const pages = [
      "This is page 1",
      "And this is page 2",
      "Now you can see page 3",
      "Look, it's page 4",
      "And finally, this is page 5",
    ];

    const StyledRoot = styled("div")({
      width: "100%",
      justifyContent: "center",
    });

    const StyledPage = styled("div")({
      textAlign: "center",
    });

    const StyledDotPagination = styled(HvDotPagination)({
      [`& .${dotPaginationClasses.radioRoot}`]: {
        marginLeft: "0px",
      },

      [`& .${dotPaginationClasses.radio}`]: {
        height: "32px",

        "&:hover": {
          backgroundColor: theme.colors.atmo3,
          borderRadius: 0,
        },
      },
    });

    return (
      <StyledRoot>
        <StyledPage>
          <HvTypography>{pages[page]}</HvTypography>
        </StyledPage>
        <br />
        <StyledDotPagination
          unselectedIcon={<RadioButtonUnselected iconSize="XS" />}
          selectedIcon={<CurrentStep iconSize="XS" />}
          page={page}
          pages={pages.length}
          onPageChange={(_, value) => setPage(value)}
          getItemAriaLabel={(pageNumber) => {
            switch (pageNumber) {
              case 0:
                return "first page button aria-label";
              case 4:
                return "last page button aria-label";
              default:
                return `${pageNumber + 1} page aria-label`;
            }
          }}
          role="navigation"
          aria-label="Example Dot Navigation"
        />
      </StyledRoot>
    );
  },
};
