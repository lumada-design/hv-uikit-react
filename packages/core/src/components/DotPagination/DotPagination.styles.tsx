import styled from "@emotion/styled";
import { HvRadio, HvRadioGroup } from "components";
import { theme } from "@hitachivantara/uikit-styles";
import dotPaginationClasses from "./dotPaginationClasses";
import { CurrentStep, OtherStep } from "@hitachivantara/uikit-react-icons";

export const StyledRadioGroup = styled(HvRadioGroup)({
  display: "flex",
  justifyContent: "center",

  [`& .${dotPaginationClasses.horizontal}`]: {
    marginLeft: 0,
    width: "auto",
  },
});

export const StyledRadio = styled((props) => <HvRadio {...props} />)({
  [`&.${dotPaginationClasses.radioRoot}`]: {
    marginLeft: "8px",
  },

  [`& .${dotPaginationClasses.radio}`]: {
    height: "16px",

    "&:hover": {
      backgroundColor: theme.colors.sema7,
      borderRadius: "100%",
    },
  },
});

export const StyledOtherStep = styled(OtherStep)({
  width: "16px",
  height: "16px",
});

export const StyledCurrentStep = styled(CurrentStep)({
  width: "16px",
  height: "16px",
});
