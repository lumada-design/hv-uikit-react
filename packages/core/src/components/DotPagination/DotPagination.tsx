import { cloneElement } from "react";
import { CurrentStep, OtherStep } from "@hitachivantara/uikit-react-icons";

import { HvRadio } from "@core/components/Radio";
import { HvRadioGroup, HvRadioGroupProps } from "@core/components/RadioGroup";
import { ExtractNames } from "@core/utils/classes";

import { useDefaultProps } from "@core/hooks/useDefaultProps";

import { staticClasses, useClasses } from "./DotPagination.styles";

export { staticClasses as dotPaginationClasses };

export type HvDotPaginationClasses = ExtractNames<typeof useClasses>;

export interface HvDotPaginationProps
  extends Omit<HvRadioGroupProps, "classes"> {
  /**
   * Icon to override the default one used for the unselected state.
   *
   * The default icon is `OtherStep`.
   */
  unselectedIcon?: React.ReactElement;
  /**
   * Icon to override the default one used for the selected state.
   *
   * The default icon is `CurrentStep`.
   */
  selectedIcon?: React.ReactElement;
  /**
   *  The number of pages the component has.
   */
  pages?: number;
  /**
   * The currently selected page.
   */
  page?: number;
  /**
   * The callback fired when the page value changes.
   */
  onPageChange?: (
    event: React.ChangeEvent<HTMLInputElement>,
    page: number
  ) => void;
  /**
   * The callback fired to get the page aria label.
   */
  getItemAriaLabel?: (page: number) => string;
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes?: HvDotPaginationClasses;
}

const getSelectorIcons = (
  radioIcon?: React.ReactElement,
  radioCheckedIcon?: React.ReactElement,
  classes?: HvDotPaginationClasses
) => {
  return {
    radio: cloneElement(radioIcon || <OtherStep width={8} height={8} />, {
      className: classes?.icon,
    }),
    radioChecked: cloneElement(
      radioCheckedIcon || <CurrentStep width={8} height={8} />,
      {
        className: classes?.icon,
      }
    ),
  };
};

const range = (n: number) => Array.from(Array(n), (_, i) => i);

/**
 * Pagination is the process of dividing a document into discrete pages. It relates to how users interact with structured content on a website or application.
 * This component uses Radio Buttons to represent each page.
 */
export const HvDotPagination = (props: HvDotPaginationProps) => {
  const {
    className,
    classes: classesProp,
    unselectedIcon,
    selectedIcon,
    pages = 1,
    page = 0,
    onPageChange,
    getItemAriaLabel,
    ...others
  } = useDefaultProps("HvDotPagination", props);

  const { classes, cx } = useClasses(classesProp);

  const icons = getSelectorIcons(unselectedIcon, selectedIcon, classes);

  return (
    <HvRadioGroup
      className={cx(classes.root, className)}
      classes={{
        horizontal: classes.horizontal,
      }}
      orientation="horizontal"
      {...others}
    >
      {range(pages).map((i) => (
        <HvRadio
          classes={{
            radio: classes.radio,
            root: classes.radioRoot,
          }}
          key={i}
          value={i}
          checked={page === i}
          onChange={(event) => onPageChange?.(event, i)}
          icon={icons.radio}
          checkedIcon={icons.radioChecked}
          aria-label={getItemAriaLabel?.(i)}
        />
      ))}
    </HvRadioGroup>
  );
};
