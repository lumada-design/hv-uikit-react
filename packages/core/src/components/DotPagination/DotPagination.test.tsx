import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { HvTypography } from "@core/components";
import { useState } from "react";
import { describe, expect, it } from "vitest";
import { HvDotPagination } from "./DotPagination";

const Pagination = ({ page }: { page: number }) => (
  <HvDotPagination
    page={page}
    pages={3}
    getItemAriaLabel={(pageNumber) => {
      switch (pageNumber) {
        case 0:
          return "first page button aria-label";
        case 2:
          return "last page button aria-label";
        default:
          return `${pageNumber + 1} page aria-label`;
      }
    }}
    role="navigation"
  />
);

const PaginationWithState = () => {
  const [page, setPage] = useState<number>(0);
  const pages = ["Page 1", "Page 2", "Page 3", "Page 4", "Page 5"];

  return (
    <div>
      <div>
        <HvTypography>{pages[page]}</HvTypography>
      </div>
      <br />
      <HvDotPagination
        page={page}
        pages={pages.length}
        onPageChange={(_, value) => setPage(value)}
        getItemAriaLabel={(pageNumber) => {
          switch (pageNumber) {
            case 0:
              return "first page button aria-label";
            case 4:
              return "last page button aria-label";
            default:
              return `${pageNumber + 1} page aria-label`;
          }
        }}
        role="navigation"
        aria-label="Example Dot Navigation"
      />
    </div>
  );
};

describe("DotPagination", () => {
  it("should have dot navigation on the first page", async () => {
    const { getByLabelText, findByRole } = render(<Pagination page={0} />);

    const dotPagination = await findByRole("navigation");

    expect(dotPagination).toBeInTheDocument();

    // Check if all buttons are present, have the right aria label and get their reference
    const radioButton1 = getByLabelText("first page button aria-label");
    const radioButton2 = getByLabelText("2 page aria-label");
    const radioButton3 = getByLabelText("last page button aria-label");

    expect(radioButton1).toBeInTheDocument();
    expect(radioButton2).toBeInTheDocument();
    expect(radioButton3).toBeInTheDocument();

    expect(radioButton1).toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).not.toBeChecked();
  });

  it("should have dot navigation on the second page", async () => {
    const { getByLabelText, findByRole } = render(<Pagination page={1} />);

    const dotPagination = await findByRole("navigation");

    expect(dotPagination).toBeInTheDocument();

    // Check if all buttons are present, have the right aria label and get their reference
    const radioButton1 = getByLabelText("first page button aria-label");
    const radioButton2 = getByLabelText("2 page aria-label");
    const radioButton3 = getByLabelText("last page button aria-label");

    expect(radioButton1).toBeInTheDocument();
    expect(radioButton2).toBeInTheDocument();
    expect(radioButton3).toBeInTheDocument();

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).toBeChecked();
    expect(radioButton3).not.toBeChecked();
  });

  it("should have dot navigation on the last page", async () => {
    const { getByLabelText, findByRole } = render(<Pagination page={2} />);

    const dotPagination = await findByRole("navigation");

    expect(dotPagination).toBeInTheDocument();

    // Check if all buttons are present, have the right aria label and get their reference
    const radioButton1 = getByLabelText("first page button aria-label");
    const radioButton2 = getByLabelText("2 page aria-label");
    const radioButton3 = getByLabelText("last page button aria-label");

    expect(radioButton1).toBeInTheDocument();
    expect(radioButton2).toBeInTheDocument();
    expect(radioButton3).toBeInTheDocument();

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).toBeChecked();
  });

  it("should have dot navigation on all pages", async () => {
    const { getByLabelText, findByRole, findByText } = render(
      <PaginationWithState />
    );

    const dotPagination = await findByRole("navigation");

    expect(dotPagination).toBeInTheDocument();

    // Check if all buttons are present, have the right aria label and get their reference
    const radioButton1 = getByLabelText("first page button aria-label");
    const radioButton2 = getByLabelText("2 page aria-label");
    const radioButton3 = getByLabelText("3 page aria-label");
    const radioButton4 = getByLabelText("4 page aria-label");
    const radioButton5 = getByLabelText("last page button aria-label");

    expect(radioButton1).toBeInTheDocument();
    expect(radioButton2).toBeInTheDocument();
    expect(radioButton3).toBeInTheDocument();
    expect(radioButton4).toBeInTheDocument();
    expect(radioButton5).toBeInTheDocument();

    await userEvent.click(radioButton1);
    await findByText("Page 1");

    expect(radioButton1).toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).not.toBeChecked();
    expect(radioButton4).not.toBeChecked();
    expect(radioButton5).not.toBeChecked();

    await userEvent.click(radioButton2);
    await findByText("Page 2");

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).toBeChecked();
    expect(radioButton3).not.toBeChecked();
    expect(radioButton4).not.toBeChecked();
    expect(radioButton5).not.toBeChecked();

    await userEvent.click(radioButton3);
    await findByText("Page 3");

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).toBeChecked();
    expect(radioButton4).not.toBeChecked();
    expect(radioButton5).not.toBeChecked();

    await userEvent.click(radioButton4);
    await findByText("Page 4");

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).not.toBeChecked();
    expect(radioButton4).toBeChecked();
    expect(radioButton5).not.toBeChecked();

    await userEvent.click(radioButton5);
    await findByText("Page 5");

    expect(radioButton1).not.toBeChecked();
    expect(radioButton2).not.toBeChecked();
    expect(radioButton3).not.toBeChecked();
    expect(radioButton4).not.toBeChecked();
    expect(radioButton5).toBeChecked();
  });
});
