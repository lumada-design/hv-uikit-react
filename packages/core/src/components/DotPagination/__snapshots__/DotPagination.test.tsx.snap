// Vitest Snapshot v1

exports[`DotPagination > should render correctly 1`] = `
<div>
  <div
    class="HvDotPagination-root e1buriah3 HvRadioGroup-root css-w7urnl-StyledFormElement-StyledRadioGroup e1qvmihv2 HvFormElement-root"
  >
    <div
      class="HvRadioGroup-group HvDotPagination-horizontal HvRadioGroup-horizontal css-2e6w85-StyledGroup e1qvmihv0"
      role="navigation"
    >
      <div
        class="e1buriah2 HvRadio-root HvDotPagination-radioRoot css-esudyc-StyledHvFormElement-StyledRadio e11n8xyl2 HvFormElement-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorDefault PrivateSwitchBase-root MuiRadio-root MuiRadio-colorDefault Mui-checked MuiRadio-root MuiRadio-colorDefault HvRadio-radio HvDotPagination-radio e11n8xyl3 HvBaseRadio-root e1sbb5xr0 css-1t388db-MuiButtonBase-root-MuiRadio-root-StyledRadio-StyledHvBaseRadio"
        >
          <input
            aria-label="first page button aria-label"
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="hvradiogroup9"
            type="radio"
            value="0"
          />
          <div
            class="HvDotPagination-icon css-10burq8-StyledCurrentStep e1buriah0 css-du18qm"
            font-size="medium"
            name="CurrentStep"
          >
            <svg
              focusable="false"
              height="8"
              viewBox="0 0 16 16"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M16 8a8.024 8.024 0 0 1-8 8 8.023 8.023 0 0 1-8-8 8.023 8.023 0 0 1 8-8 8.024 8.024 0 0 1 8 8z"
                fill="var(--colors-acce1)"
              />
            </svg>
          </div>
        </span>
      </div>
      <div
        class="e1buriah2 HvRadio-root HvDotPagination-radioRoot css-esudyc-StyledHvFormElement-StyledRadio e11n8xyl2 HvFormElement-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorDefault PrivateSwitchBase-root MuiRadio-root MuiRadio-colorDefault MuiRadio-root MuiRadio-colorDefault HvRadio-radio HvDotPagination-radio e11n8xyl3 HvBaseRadio-root e1sbb5xr0 css-1t388db-MuiButtonBase-root-MuiRadio-root-StyledRadio-StyledHvBaseRadio"
        >
          <input
            aria-label="2 page aria-label"
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="hvradiogroup9"
            type="radio"
            value="1"
          />
          <div
            class="HvDotPagination-icon css-1tb1fcm-StyledOtherStep e1buriah1 css-du18qm"
            font-size="medium"
            name="OtherStep"
          >
            <svg
              focusable="false"
              height="8"
              viewBox="0 0 16 16"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M12 8a4.012 4.012 0 0 1-4 4 4.012 4.012 0 0 1-4-4 4.012 4.012 0 0 1 4-4 4.012 4.012 0 0 1 4 4z"
                fill="var(--colors-acce1)"
              />
            </svg>
          </div>
        </span>
      </div>
      <div
        class="e1buriah2 HvRadio-root HvDotPagination-radioRoot css-esudyc-StyledHvFormElement-StyledRadio e11n8xyl2 HvFormElement-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorDefault PrivateSwitchBase-root MuiRadio-root MuiRadio-colorDefault MuiRadio-root MuiRadio-colorDefault HvRadio-radio HvDotPagination-radio e11n8xyl3 HvBaseRadio-root e1sbb5xr0 css-1t388db-MuiButtonBase-root-MuiRadio-root-StyledRadio-StyledHvBaseRadio"
        >
          <input
            aria-label="last page button aria-label"
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="hvradiogroup9"
            type="radio"
            value="2"
          />
          <div
            class="HvDotPagination-icon css-1tb1fcm-StyledOtherStep e1buriah1 css-du18qm"
            font-size="medium"
            name="OtherStep"
          >
            <svg
              focusable="false"
              height="8"
              viewBox="0 0 16 16"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M12 8a4.012 4.012 0 0 1-4 4 4.012 4.012 0 0 1-4-4 4.012 4.012 0 0 1 4-4 4.012 4.012 0 0 1 4 4z"
                fill="var(--colors-acce1)"
              />
            </svg>
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;
